[IMP] sale_coupon_product_management: program vals

oerp-odoo opened the pull request:
Now program/product values generated from options can use optional
`program` record, so it could access dynamic values (in this case for
related product).

oerp-odoo committed:
[IMP] sale_coupon_product_management: program vals

Now program/product values generated from options can use optional
`program` record, so it could access dynamic values (in this case for
related product).

mmequignon commented on sale_coupon_product_management/models/sale_coupon_program_option.py:
```diff
@@ -98,7 +98,14 @@ def _get_related_record(self, program, path):
         return operator.attrgetter(path)(program)
 
-    def get_program_values(self):
+    def _update_values_from_program(self, opt_vals, program=None):
```
Since that method does nothing when program isn't set, why do you set this attribute as optional ?

oerp-odoo replied:
Because it comes from parent method, which has that record. Though I guess I could move check outside that function, to look more clean.

mmequignon commented on sale_coupon_product_management/models/sale_coupon_program_option.py:
```diff
@@ -98,7 +98,14 @@ def _get_related_record(self, program, path):
-    def get_program_values(self):
+    def _update_values_from_program(self, opt_vals, program=None):
+        self.ensure_one()
+        if program and self.option_type == "discount_fixed_amount":
+            opt_vals[DISCOUNT_PRODUCT_FNAME][
+                "list_price"
+            ] = program.discount_fixed_amount
```
It's better to return a new dict to update the dict from outer scope than modify a mutable object.

mmequignon commented on sale_coupon_product_management/models/sale_coupon_program_option.py:
```diff
@@ -114,6 +121,7 @@ def get_program_values(self):
         for option in self:
             opt_cfg = cfg[option.option_type]
             opt_vals = opt_cfg["values"]
+            option._update_values_from_program(opt_vals, program=program)
```
then
```suggestion
            opt_vals.update(option._update_values_from_program(program))
```

oerp-odoo replied:
Well in this case it would complicate things, because this is not single level dictionary. It can be multi level dictionary, which then would need to be built inside `_update_values_from_program` (using template), then updated and when returned, merged with existing opt_vals.

If I just update single value, I dont need to do any of that and main method (`get_program_values`) does not need to be even aware how it is done.

mmequignon replied:
You just have to return `{DISCOUNT_PRODUCT_FNAME: {"list_price": program.discount_fixed_amount}}`.
Or maybe there's something I don't see?

oerp-odoo replied:
You can't, because if you would use it on `update`, it would overwrite existing dict on `DISCOUNT_PRODUCT_FNAME` key.  That dict might be already prefilled with other values (thus why there is multi-level dict).

As mentioned by TODO: its possible to use `mergedeep` module that can help, but still it complicates this, cause we would then merge dictionaries twice. In this case and after.

oerp-odoo replied:
In other words you want to either merge dictionaries (not update) or update one specific value -> which in my opinion looks simpler (if we would need to update many key/values, then merging might be more beneficial).

mmequignon replied:
then you could maybe do it like this : `opt_vals[DISCOUNT_PRODUCT_FNAME].update(…)`
or just return the value to assign.
But whatever you chose, we should never modify a mutable object from a method.

oerp-odoo replied:
But then main method starts doing more then it should. its responsibility for `_update_values_from_program` and it should be able to update not just whats on `DISCOUNT_PRODUCT_FNAME` (it so happens that in this case we now update only one value there, but it could be that we would need to update more. And then it would mean multiple `update` calls)

oerp-odoo committed:
[FIX] sale_coupon_product_management: onchange

_predicate_product_categ_with_opts was called on incorrect object. Also
added test to make sure onchange now works correctly.

jcoux commented on sale_coupon_product_management/models/sale_coupon_program_option.py:
```diff
@@ -98,7 +98,14 @@ def _get_related_record(self, program, path):
+        if self.option_type == "discount_fixed_amount":
+            opt_vals[DISCOUNT_PRODUCT_FNAME][
+                "list_price"
+            ] = program.discount_fixed_amount
```
As asked by @mmequignon don't change values of `opt_vals` directly into the function.
Return the new values and update the dict out of the function.

jcoux commented on sale_coupon_product_management/models/sale_coupon_program_option.py:
```diff
@@ -114,6 +121,8 @@ def get_program_values(self):
             opt_cfg = cfg[option.option_type]
             opt_vals = opt_cfg["values"]
+            if program:  # TODO: a bit redundant.
+                option._update_values_from_program(opt_vals, program)
```
Same comment, follow the @mmequignon proposition.
Thanks.

oerp-odoo replied:
Updated, but it now made implementation more complicated and harder to extend.

Also where you get this "mutable object" should never be updated inside method? Tried to look for such antipattern, but people say its not antipattern, as long as intention/method is named clearly (like mentioning `update` word or something similar).

TBF `dict.update` that you wanted to use instead of my custom method, actually does same thing. It updates mutable object inside method/function. Looks like nitpicking to me:)

mmequignon replied:
There's plenty of examples and answers everywhere (first result https://softwareengineering.stackexchange.com/a/245809)

Regarding your comment, you're right, maybe the best place to set would be in `_get_program_values_cfg`
But there's no need to add dozen of intermediate layers where each one modify the same dict.
IMO, its the `get_program_values` responsibility to aggregate values into a single dict.

And regarding your comment in the code, it's not a bad design.

oerp-odoo replied:
OK, I can remove that comment if you think its a good design.

Though that link you put, just talks in general and people actually comment it depends on language and context. So its not black and white.

Also `update` function for dict just does what you say can't be done. So it contradicts saying that object can't be modified inside method/function (then you would not be able to update it at all).

https://stackoverflow.com/questions/58533878/change-object-directly-in-function-is-anti-pattern-in-python

And regarding my comment about it being bad design. It actually has nothing to do where update happens. It has to do with the fact, that now that helper method has become poorly extendable, because we would need to change returned structure and/or how returned dict is treated if any new values would need to be introduced to be used from `program` record or some other related record.

It also breaks SoC, because caller needs to know more than it should have (how to treat returned dict). So its now kind of hardcoded:)

oerp-odoo committed:
[REF] sale_coupon_product_management: upd elsewhere

jcoux commented:
Waiting review update from @mmequignon before update my review.

jcoux commented on sale_coupon_product_management/models/sale_coupon_program_option.py:
```diff
@@ -114,6 +120,15 @@ def get_program_values(self):
         for option in self:
             opt_cfg = cfg[option.option_type]
             opt_vals = opt_cfg["values"]
+            if program:  # TODO: a bit redundant.
+                # NOTE. This is bad design. If values from program would
+                # need to include extra values on different level dicts,
+                # this could easily break or work incorrectly. Its
+                # because get_program_values needs to be aware what
+                # _prepare_values_from_program needs to return.
+                opt_vals[DISCOUNT_PRODUCT_FNAME].update(
+                    option._prepare_values_from_program(program)
+                )
```
Agree with @mmequignon comment, it's not a bad design.
Please remove your comment and we will can approve your PR.
```suggestion
            if program:
                opt_vals[DISCOUNT_PRODUCT_FNAME].update(
                    option._prepare_values_from_program(program)
                )
```

oerp-odoo committed:
fixup! [REF] sale_coupon_product_management: upd elsewhere

oerp-odoo commented:
@jcoux @mmequignon updated

jcoux reviewed (approved):
Thanks for changes
